tt_eager/tensor/tensor.hpp

```diff
@@ -31,19 +31,21 @@ struct Tensor {
         ttnn::Shape shape;
         DataType dtype;
         Layout layout;
-        std::atomic<bool> metadata_populated = false;
+        std::mutex populated_mutex;
+        std::vector<bool> tensor_populated = {};
         uint32_t main_thread_ref_count = 0;
         bool deallocated = false; // Set to true if device side storage was deallocated
         bool dynamic_storage = false; // Storage type can change, depending on op behaviour
+        bool is_borrowed = false; // Set to true if the tensor is using borrowed storage. Need to copy the storage if using async mode.
```

Comment:
why do you need this attribute? can't you figure this out from the StorageType?

Reply:
The goal here was to determine if we're using a borrowed buffer without querying the storage type in the main thread. This query can require the main thread to block until the worker has populated the tensor.

Only the main thread sets and queries this value, minimizing interactions between threads, which causes less stalls in certain cases.

Reply:
For multi-device, we only allow OwnedStorage as the underlying storagetype for the MultiDeviceHostStorage. Borrowed would only happen if you're on single-device. Does that simplify things?

Reply:
You're right about owned storage being the one used for `MultiDeviceHostStorage`. We need this for single device use-cases running in async mode, where a user chooses to directly call a ttl operation on a torch tensor.

I think the confusion is arising from me having added this feature along with multi-device changes. It should haave been added separately but I was trying to bundle in this change since it doesn't have an assert on main which can cause async tests to segfault if running with borrowed storage,

Reply:
But if it's only used for single-device, there's only the main thread and the worker dispatch so the overhead shouldn't really be a concern here?

Reply:
Yes, we only have async mode tests using owned storage. To your point about this not being a concern wrt single device: I think having the main thread unnecessarily block even when we have a single worker is not ideal. Additionally, we have shared functions for both single and multi-device workload, in which case we end up doing the storage type check for multi-device as well.

If you guys think its significantly cleaner to just infer the buffer type from the storage, I can add the change to this PR.

Reply:
Based on an offline discussion with Joseph, I've removed this attribute for now. If we need it in future as part of optimization passes, we can add it back.